gkctl/main.c

```diff
@@ -0,0 +1,218 @@
+		}
+		args->lua_script_path = arg;
+		break;
+
+	case ARGP_KEY_END:
+		if (!args->lua_script_path) {
+			argp_error(state,
+				"The lua script path was not specified");
+		}
+		break;
+
+	default:
+		return ARGP_ERR_UNKNOWN;
+	}
+
+	return 0;
+}
+
+static struct argp argp = {options, parse_opt, adoc, doc, NULL, NULL, NULL};
+
+static int
+load_file_to_buffer(const char *file_name, char *buffer, int n)
+{
+	int ret = 0;
+	FILE *fp;
+
+	memset(buffer, 0, n);
+
+	if ((fp = fopen(file_name, "r")) == NULL) {
+		fprintf(stderr, "Error: Failed to open file %s - %s\n",
+			file_name, strerror(errno));
+		ret = -1;
+		goto out;
+	}
+
+	/*
+	 * Return value equals the number of bytes transferred
+	 * only when size (i.e., second parameter) is 1.
+	 */
+	ret = fread(buffer, 1, n, fp);
+	if ((ret < n) && ferror(fp)) {
+		fprintf(stderr, "Error: Failed to read file %s - %s\n",
+			file_name, strerror(errno));
+		ret = -1;
+	}
+
+	fclose(fp);
+out:
+	return ret;
+}
+
+int
+main(int argc, char *argv[])
+{
+	int ret = 0;
+	int sock_fd;
+	char *line = NULL;
+	char file_str[MSG_MAX_LEN];
+	char send_buff[MSG_MAX_LEN + 2];
+	char recv_buff[MSG_MAX_LEN + 2];
+	size_t len = 0;
+	size_t total_file_len = 0;
+	uint16_t msg_len;
+	int send_size = 0;
+	int recv_size = 0;
+	int tot_send = 0;
+	int tot_recv;
+	struct sockaddr_un serv_addr; 
+
+    	struct args args = {
+		/* Defaults. */
+		.server_path = "/tmp/dyn_cfg.socket",
+	};
+
+	/* Read parameters. */
+	argp_parse(&argp, argc, argv, 0, NULL, &args);
+
+	if (sizeof(serv_addr.sun_path) < strlen(args.server_path)) {
+		fprintf(stderr, "Error: passing a too long server path - %s!\n",
+			args.server_path);
+		ret = -1;
+		goto out;
+	}
+
+	ret = load_file_to_buffer(args.lua_script_path,
+		file_str, sizeof(file_str));
+	if (ret <= 0) {
+		ret = -1;
+		goto out;
+	}
+	total_file_len = ret;
+	msg_len = htons(ret);
+
+	memcpy(send_buff, &msg_len, sizeof(msg_len));
+	memcpy(send_buff + sizeof(msg_len), file_str, total_file_len);
+
+	if((sock_fd = socket(AF_UNIX, SOCK_STREAM, 0)) < 0) {
+		perror("Error : Could not create socket\n");
+		ret = -1;
+		goto out;
+	}
+
+	memset(&serv_addr, 0, sizeof(serv_addr));
+	serv_addr.sun_family = AF_UNIX;
+	strncpy(serv_addr.sun_path, args.server_path,
+		strlen(args.server_path));
+
+	if(connect(sock_fd, (struct sockaddr *)&serv_addr,
+			sizeof(serv_addr)) < 0) {
+		perror("Error : Connect failed\n");
+		ret = -1;
+		goto close_sock;
+	}
+
+	len = total_file_len + sizeof(uint16_t);
+	while ((send_size = send(sock_fd, send_buff + tot_send,
+			len - tot_send, 0)) > 0) {
+		tot_send += send_size;
+		if (tot_send >= len)
+			break;
+	}
+
+	if (tot_send != len) {
+		fprintf(stderr, "Send incomplete message!\n");
+		ret = -1;
+		goto close_sock;
+	}
+
+	memset(recv_buff, 0,sizeof(recv_buff));
+	if ((recv_size = recv(sock_fd, recv_buff, sizeof(recv_buff), 0)) < 0) {
+		perror("Error : Receive failed\n");
+		ret = -1;
+		goto close_sock;
+	}
+	tot_recv = ntohs(*(uint16_t *)recv_buff);
```

Comment:
I understand that there's no guarantee that the message will arrive all at once.

Reply:
Yes, it's possible.